pyscf: 1.7.6.post1 -> 2.0.1

sheepforce opened the pull request:
###### Motivation for this change
This updates PySCF to its latest release. The build system has changed somewhat and now uses a lot of CMake. I've enabled the tests, as inspired by an older commit of @drewrisinger .
The CPPE library for polarisable embedding has been added, which is an optional dependency of PySCF. I am not entirely sure if i did the python binding stuff correctly there. CPPE is at least not in the `PYTHONPATH` from within a nix-shell.

###### Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jonringer commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1749,6 +1749,8 @@ in {
 
   cozy = callPackage ../development/python-modules/cozy { };
 
+  cppe = toPythonModule (pkgs.cppe.override { enablePython = true; });
```
The python version also needs to be set, or else you will get `pkgs.python3` for all interpreter versions
```suggestion
  cppe = toPythonModule (pkgs.cppe.override {
    enablePython = true;
    python3 = python;
  });
```

sheepforce replied:
I've passed python explicitly, however the PySCF `pythonCatchConflictsPhase` now complains about duplicate dependencies in the closure (pybind, numpy, scipy, h5py and six), coming in directly as raw python packages, and wrapped into a python env (I think from `python.withPackages`). I'm now passing all Python packages directly, as nghttp2 does it [here](https://github.com/NixOS/nixpkgs/blob/b67e752c29f18a0ca5534a07661366d6a2c2e649/pkgs/development/libraries/nghttp2/default.nix#L12), and this reduces the conflicts to `pybind11`. But still there is the duplicate and I am not sure where this comes from. Any idea?

markuskowa commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,48 @@
+, python3, enablePython ? false
+}:
+
+let
```
This let could go directly to the `cmakeFlags`?

markuskowa commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,48 @@
+  propagatedBuildInputs = lib.optional enablePython pythonWP;
+
+  #pythonSitePackages = "lib/${pythonWP.executable}/site-packages";
+  #setupHooks = lib.optional enablePython ./pythonBindingsHook.sh;
```
These lines can be removed?

sheepforce commented:
As written I've enabled the test suite. However, it is huge and runs for at least 2 h on my 10 cores and some single test are failing and need to be disabled. I guess it might be better to disable the full test suite? What do you think?

markuskowa commented:
> As written I've enabled the test suite. However, it is huge and runs for at least 2 h on my 10 cores and some single test are failing and need to be disabled. I guess it might be better to disable the full test suite? What do you think?

Let's disable the test suite for now. It is too big and not reliable. We still have a test in NixOS-QChem to catch basic errors.

SuperSandro2000 commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,45 @@
+}:
+
+let
+  pythonWP = python3.withPackages (p: with p; [
```
```suggestion
  pythonEnv = python3.withPackages (p: with p; [
```
or simply py

sheepforce replied:
Changed in favour of passing all python packages directly, unfortunately still producing conflicts, see [here](https://github.com/NixOS/nixpkgs/pull/144253#discussion_r741921129)

sheepforce replied:
Followed the approach in https://github.com/NixOS/nixpkgs/pull/144253#discussion_r742069051, which solves the problems with much less hacks.

drewrisinger commented:
> 
> 
> As written I've enabled the test suite. However, it is huge and runs for at least 2 h on my 10 cores and some single test are failing and need to be disabled. I guess it might be better to disable the full test suite? What do you think?

Even PySCF mainline doesn't run all tests that they include in their source: https://github.com/pyscf/pyscf/blob/master/.github/workflows/run_tests.sh

Here are the tests that I excluded, feel free to use it as a starting point. I think tests typically ran for about 20-30 mins with all these disabled? https://github.com/drewrisinger/nur-packages/blob/77d8267ff1b432c76ae90fc57db9e9787640137f/pkgs/python-modules/pyscf/default.nix#L94-L255

drewrisinger commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,44 @@
+    numba
+    scipy
+    pandas
+    pytest
```
This probably shouldn't be a ``propagatedBuildInput``, I'd expect it in ``checkInputs`` (and maybe use ``pytestCheckHook`` instead for auto-testing)

SuperSandro2000 replied:
propagating pytest is a big no no because it makes it really hard to get any other version for checkPhase.

drewrisinger commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,44 @@
+{ stdenv, lib, fetchFromGitHub, cmake
```
Instead of making this file generate both the Python bindings and the regular C bindings, maybe you'd consider following the approach that I used in my NUR with the similar (in terms of packaging) ``tweedledum`` library:
* 1 Nix module for the C library: https://github.com/drewrisinger/nur-packages/blob/8c70c747b48bd8b13eb3173aa27d6afdb7f694b1/pkgs/libraries/tweedledum/default.nix
* 1 Nix module for the python library, which inherits the source from the C library: https://github.com/drewrisinger/nur-packages/blob/8c70c747b48bd8b13eb3173aa27d6afdb7f694b1/pkgs/python-modules/tweedledum/default.nix

It's slightly more Nix packaging, but I doubt you'd run into any python version conflicts this way.

sheepforce replied:
This was a very good solution. Using the setup.py build solved all the python problems.

drewrisinger commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,44 @@
+    owner = "maxscheurer";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "guM7+ZWDJLcAUJtPkKLvC4LYSA2eBvER7cgwPZ7FxHw=";
```
See other comment on SRI hashes

drewrisinger commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,21 +1,26 @@
-    rev = "f6c9c6654dd9609c5e467a1edd5c2c076f793acc";
-    sha256  = "0xbwkjxxysfpqz72qn6n4a0zr2h6sprbcal8j7kzymh7swjy117w";
+    rev = "v${version}";
+    sha256  = "J6t/Oq7WE5wz2aefok5c47CeiP2DskLF//JWGNzgRpw=";
```
This looks like a base64-encoded sha256. Would probably be worth changing to the SRI rep: ``hash = "sha256-..."`

drewrisinger commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,21 +1,26 @@
+  # setup.py calls Cmake and passes the arguments in CMAKE_CONFIGURE_ARGS to cmake.
+  nativeBuildInputs = [ cmake ];
+  dontUseCmakeConfigure = true;
+  CMAKE_CONFIGURE_ARGS = toString [
```
Current best practice that I've seen is NOT to have environment variables in the Nix module, but export this var in ``preBuild``. I think it's so that this variable doesn't leak to parts of the build process where it's not needed.

drewrisinger reviewed (changes requested):
General comments about structure of cppe derivation, and a few details on the hash types & build variables

drewrisinger commented:
> 
> 
> > As written I've enabled the test suite. However, it is huge and runs for at least 2 h on my 10 cores and some single test are failing and need to be disabled. I guess it might be better to disable the full test suite? What do you think?
> 
> Let's disable the test suite for now. It is too big and not reliable. We still have a test in NixOS-QChem to catch basic errors.

That might be true, but the issue is that any packaging errors should ideally be caught before merge, not after. Do you expect a potential reviewer of future ``pyscf`` updates to know to go to NixOS-QChem, figure out how to run tests there against a PR, and then review? Vs the standard nixpkgs review process of ``nix-review pr ...``?

SuperSandro2000 commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,44 @@
+{ stdenv, lib, fetchFromGitHub, cmake
+, python, pybind11 ? null, numpy ? null, h5py ? null, numba ? null, scipy ? null
+, pandas ? null, pytest ? null, enablePython ? false
```
Having ``? null`` is a big anti pattern.

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1749,6 +1749,11 @@ in {
   cozy = callPackage ../development/python-modules/cozy { };
 
+  cppe = toPythonModule (pkgs.cppe.override {
+    inherit (self) python pybind11 numpy h5py numba scipy pandas pytest;
```
This shouldn't be required. I think we can input python and use with python in the package.

markuskowa commented:
> > > As written I've enabled the test suite. However, it is huge and runs for at least 2 h on my 10 cores and some single test are failing and need to be disabled. I guess it might be better to disable the full test suite? What do you think?
> > 
> > 
> > Let's disable the test suite for now. It is too big and not reliable. We still have a test in NixOS-QChem to catch basic errors.
> 
> That might be true, but the issue is that any packaging errors should ideally be caught before merge, not after. Do you expect a potential reviewer of future `pyscf` updates to know to go to NixOS-QChem, figure out how to run tests there against a PR, and then review? Vs the standard nixpkgs review process of `nix-review pr ...`?

I agree with you, but unreliable test suites are worse than (or as good as) not having a test at all. I would not expect that some reviewer uses an external test. I look at the NixOS-QChem tests as a fallback option, which catches the error at some point later.

sheepforce commented:
Thank you for the valuable feedback @drewrisinger , this solved some problems. I am doing final tunings on the test suite to get it running properly, and then it should be done.

drewrisinger commented:
FWIW, release 2.0.1 was posted a few hours ago. Might want to roll that into this PR if possible, since it's still not finalized? https://github.com/pyscf/pyscf/releases/tag/v2.0.1

drewrisinger commented on pkgs/development/python-modules/cppe/default.nix:
```diff
@@ -0,0 +1,35 @@
+      --replace '"external/eigen3",' '"${eigen}/include/eigen3",'
+  '';
+
+  nativeBuildInputs = [ cmake pytestCheckHook ];
```
I'd change ``pytestCheckHook`` -> a ``checkInput``

sheepforce replied:
Not entirely sure what you mean. Do you mean putting `pytest` in `checkInput`? This one does not work with `disabledTestPaths` then.

sheepforce replied:
Nevermind, figured out what you've meant :smile:

drewrisinger commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -28,12 +31,57 @@ buildPythonPackage rec {
 
+  checkInputs = [ nose nose-exclude ];
+
   PYSCF_INC_DIR="${libcint}:${libxc}:${xcfun}";
```
```suggestion
  PYSCF_INC_DIR="${libcint}:${libxc}:${xcfun}";
```
I think this should be a ``preBuild`` export?

sheepforce replied:
Already in the `preConfigure` but yes :)

drewrisinger commented on pkgs/development/python-modules/cppe/default.nix:
```diff
@@ -0,0 +1,35 @@
+}:
+
+buildPythonPackage rec {
+  inherit (cppe) pname version src;
```
```suggestion
  inherit (cppe) pname version src meta;
```

drewrisinger commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -28,12 +31,57 @@ buildPythonPackage rec {
+    runHook preCheck
+
+    nosetests pyscf/ -v \
+        --exclude-dir=examples --exclude-dir=pyscf/pbc/grad \
```
(optional): these excludes can be passed as environment variables, e.g. ``NOSE_EXCLUDE=["test_bz", "h2o_vdz", ...];``

This allows commenting the individual tests better (i.e. this is actually broken, this is slow, etc.)

Env var names/options correlation can be found here: https://nose.readthedocs.io/en/latest/usage.html#options

sheepforce replied:
Yes, I've tried this version as in your old version of the derivation. I am not sure why, but it didn't have any effect, whether I put them only in the derivation directly, as your old code did, or export them in the `preCheck` explicitly.

drewrisinger replied:
Did you have the ``nose-exclude`` plugin included on the path (``checkInputs``)? If so, weird. I don't know why that would be.

sheepforce replied:
Yes, was available :open_mouth: I am also a little bit lost there, as I can't find the difference to you version. Anyway, I'm approaching an end with getting the test suite to run.

drewrisinger reviewed (changes requested):
Few minor suggestions

SuperSandro2000 commented on pkgs/top-level/python-packages.nix:
```diff
@@ -1751,6 +1751,8 @@ in {
 
   cozy = callPackage ../development/python-modules/cozy { };
 
+  cppe = callPackage ../development/python-modules/cppe { cppe = pkgs.cppe; };
```
```suggestion
  cppe = callPackage ../development/python-modules/cppe {
    cppe = pkgs.cppe;
  };
```

SuperSandro2000 commented on pkgs/development/python-modules/pyscf/default.nix:
```diff
@@ -1,21 +1,25 @@
-    rev = "f6c9c6654dd9609c5e467a1edd5c2c076f793acc";
-    sha256  = "0xbwkjxxysfpqz72qn6n4a0zr2h6sprbcal8j7kzymh7swjy117w";
+    rev = "v${version}";
+    hash  = "sha256-nwnhaqSn/9WHBjUPaEabK4x23fJ83WwEYvz6aCcvsDw=";
```
```suggestion
    sha256 = "sha256-nwnhaqSn/9WHBjUPaEabK4x23fJ83WwEYvz6aCcvsDw=";
```

SuperSandro2000 commented on pkgs/development/python-modules/cppe/default.nix:
```diff
@@ -0,0 +1,36 @@
+  nativeBuildInputs = [ cmake ];
+  dontUseCmakeConfigure = true;
+
+  buildInputs = [ pybind11 eigen ];
```
Is a binary from eigen called?

sheepforce replied:
No, it just uses its header files. You mean better move it to `nativeBuildInputs`?

SuperSandro2000 commented on pkgs/development/python-modules/cppe/default.nix:
```diff
@@ -0,0 +1,36 @@
+
+  pythonImportsCheck = [ "cppe" ];
+
+  # Tests that require "polarizationsolver", which is not available.
```
Can we package that easily?

sheepforce replied:
Probably yes, I will have a look: https://gitlab.com/reinholdt/polarizationsolver
There is no release for this package, but looks easy otherwise. It is just used for the tests, although and nowhere else.

sheepforce replied:
I've added `polarizationsolver` and its dependency `fields` and no tests in `cppe` need to be disabled anymore.

SuperSandro2000 commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,25 @@
+  meta = with lib; {
+    description = "C++ and Python library for Polarizable Embedding";
+    homepage = "https://github.com/maxscheurer/cppe";
+    license = licenses.lgpl3;
```
lgpl3Only or lgpl3Plus?

drewrisinger commented on pkgs/development/libraries/libcint/default.nix:
```diff
@@ -9,21 +9,25 @@
   cmakeFlags = [
     "-DENABLE_TEST=1"
     "-DQUICK_TEST=1"
+
```
Extra new-line

drewrisinger commented on pkgs/development/python-modules/cppe/default.nix:
```diff
@@ -0,0 +1,36 @@
+{ buildPythonPackage, lib, cmake, cppe, eigen
+, python, pybind11, numpy, h5py, numba, scipy
+, pandas, polarizationsolver, pytestCheckHook
+}:
```
suggest each on an individual new line

sheepforce replied:
Ran nixpkgs-fmt on all and sorted

drewrisinger commented on pkgs/development/python-modules/cppe/default.nix:
```diff
@@ -0,0 +1,36 @@
+  # Using the nixpkgs version instead.
+  postPatch = ''
+    substituteInPlace setup.py \
+      --replace '"external/eigen3",' '"${eigen}/include/eigen3",'
```
```suggestion
      --replace 'external/eigen3' '${eigen}/include/eigen3'
```
remove maybe-unnecessary quotes (didn't check the setup.py to see if this actually works)

drewrisinger commented on pkgs/development/python-modules/cppe/default.nix:
```diff
@@ -0,0 +1,36 @@
+  buildInputs = [ pybind11 ];
+
+  checkInputs = [
+    numpy
+    h5py
+    numba
+    scipy
+    pandas
+    polarizationsolver
+    pytestCheckHook
```
I generally prefer to sort these alphabetically, with ``pytestCheckHook`` at the top.

drewrisinger commented on pkgs/development/python-modules/polarizationsolver/default.nix:
```diff
@@ -0,0 +1,34 @@
+{ buildPythonPackage, lib, fetchFromGitLab, python
+, numpy, scipy, periodictable, fields
```
divide into one line per argument

drewrisinger commented on pkgs/development/python-modules/polarizationsolver/default.nix:
```diff
@@ -0,0 +1,34 @@
+
+buildPythonPackage rec {
+  pname = "polarizationsolver";
+  version = "00424ac4";
```
best practice I've seen with this is have

``version = "unstable-2021-XX-XX"`` (i.e. the date of the commit), and specify the full git hash revision in ``src.rev``

drewrisinger commented on pkgs/development/python-modules/polarizationsolver/default.nix:
```diff
@@ -0,0 +1,34 @@
+    sha256 = "sha256-LACf8Xw+o/uJ3+PD/DE/o7nwKY7fv3NyYbpjCrTTnBU=";
+  };
+
+  propagatedBuildInputs = [
+    numpy
+    scipy
+    periodictable
+  ];
```
sort alphabetically

sheepforce committed:
libcint: 4.4.0 -> 4.4.6


libcint: formatting and features


libcint: platforms

sheepforce commented on pkgs/development/libraries/science/chemistry/cppe/default.nix:
```diff
@@ -0,0 +1,25 @@
+    description = "C++ and Python library for Polarizable Embedding";
+    homepage = "https://github.com/maxscheurer/cppe";
+    license = licenses.lgpl3Only;
+    platforms = platforms.linux;
```
Not tested and there seems to be no clear rule how to deal with this? I've changed to `[ "x86_64-linux" "x86_64-darwin" ]` for all, but it would likely also run with `aarch64` stuff (at least I don't see any reason why it shouldn't. What do you think?

sheepforce replied:
Will resolve the others on favour of this and change them according to what we figure out here.

sheepforce replied:
Alrighty, changed it to `platforms.unix`, as there is no reason it would be tied to specific architecture.

sheepforce committed:
libxc: force 3rd and 4th derivatives compilation


libxc: formatting


libxc: platforms

sheepforce committed:
cppe: init at 0.3.1


cppe: move pytestCheckHook to checkInputs


cppe: hash


cppe: license and hash


cppe: formatting


python3.pkgs.cppe: more tests


cppe: formatting


cppe: formatting


cppe: platforms


cppe: platforms

sheepforce commented:
Some changes to libcint and libxc were necessary to fix some exotic DFT tests and stuff like periodic coupled cluster. The absolute majority of the test suite is running. I've just disabled a few numerically problematic tests, that do not converge to the tested 10E-11 hartree, but to a few 10E-10 hartree, or tests that even themselves print a warning, that the FFT is numerically instable.

@markuskowa , @drewrisinger , please have a final look. From my perspective, everything should be fine now.

markuskowa reviewed (commented):
This test fails for me on an AMD Epyc machine: `FAIL: test_n3_cis_ewald (test_ci.KnownValues)`.
I also noticed that the build of libxc increased significantly.

sheepforce commented:
> This test fails for me on an AMD Epyc machine: `FAIL: test_n3_cis_ewald (test_ci.KnownValues)`.

Naw, that's not so good :disappointed: Can confirm this also happens reproducibly on my i7-7700K, but it works perfectly fine on a Skylake Xeon-W at work. I wonder if we run into something like #59708 again. I will build with MKL just for testing and see if this changes anything. Otherwise I have no better idea than disabling this test and open a bug upstream?



> I also noticed that the build of libxc increased significantly.
Yes, this is unfortunately the price to pay for higher order derivatives, that PySCF requires. By default LibXC builds with first and second derivatives, but I had to enable at least the 3rd derivatives.

sheepforce commented:
OK, without changing anything and just forcing the rebuild, some other test now fail with numerical instabilities (huge ones, 10e92 hartree difference).

sheepforce commented:
Using MKL leads to numerical problems already in the dependencies; `python3.pkgs.tables` does not pass its test suite anymore.

drewrisinger reviewed (changes requested):
* [x] Diff mostly LGTM, few nitpicks
* [x] Commits LGTM
* [x] Checked licenses for new packages
* [ ] Build via ``nix-review``: see errors below

I get a few errors when building ``pyscf``, even when single-threaded. The most alarming is
``
...
File "/build/source/pyscf/gto/basis/parse_nwchem.py", line 219, in search_seg
  with open(basisfile, 'r') as fin:
OSError: [errno 24] Too many open files '/build/source/pyscf/gto/basis/ano.dat'

Rank 1820 tests in 2536.987 s
FAILED (SKIP=4, errors=6)
```

drewrisinger commented on pkgs/development/libraries/libcint/default.nix:
```diff
@@ -9,13 +9,13 @@
 stdenv.mkDerivation rec {
   pname = "libcint";
-  version = "4.4.0";
+  version = "4.4.6";
```
EDIT: I think it does, at least on my NUR repo w/ pyscf==1.7.6

sheepforce replied:
Upstream builds this release against 4.4.6 at least, as far as I can see. I would leave it for now as is to not make it even more difficult and maybe make another PR later. :slightly_smiling_face:

sheepforce commented on pkgs/development/python-modules/polarizationsolver/default.nix:
```diff
@@ -0,0 +1,39 @@
+  meta = with lib; {
+    description = "Multipole moment solver for quantum chemistry and polarisable embedding";
+    homepage = "https://gitlab.com/reinholdt/polarizationsolver";
+    license = licenses.gpl3Only;
```
You're right. Changed

sheepforce committed:
pyscf: 1.7.6.post1 -> 2.0.1


pyscf: hash


pyscf: limit test suite to single core


pyscf: adapting test suite


pyscf: fix pythonpath for tests


pyscf: formatting


pyscf: platforms


remove log


pyscf: enable uadc module


pyscf: platforms


pyscf: formatting


pyscf: disable instable N3 CI test


pyscf: formating


pyscf: increase ulimit


pyscf: ulimit files


pyscf: remove ulimit -n

sheepforce commented:
> I get a few errors when building `pyscf`, even when single-threaded. The most alarming is `` ... File "/build/source/pyscf/gto/basis/parse_nwchem.py", line 219, in search_seg with open(basisfile, 'r') as fin: OSError: [errno 24] Too many open files '/build/source/pyscf/gto/basis/ano.dat'
> 
> Rank 1820 tests in 2536.987 s FAILED (SKIP=4, errors=6)

I am afraid I cannot reproduce any of those :neutral_face: I have it running now on all of my three different CPUs (i7-7700K, Xeon W2155 and the old Xeon E5420). Unfortunately I do not have access to others, such as the AMD Epyc machines of @markuskowa . But in general numerical effects this large are worrying. However, I don't see what we could change about it, except disabling those tests.

I am forcing the test to be single core always and the only parallel part is BLAS and running them multi-core gives even less stable results.

Regarding your failing NWChem parser test; I suppose this might actually depend on how many other files you have open during your build? I don't think we can change this. `ulimit -n $LARGENUMBER` is not allowed and the number of open file handles would need to be set in the sysctl configuration. I've increased this number on my machines from the defaults some time ago.

Upstream builds with netlib BLAS in the CI apparently. I mean we could force `blas` to be `blas-reference`, but I guess this it not the way forward ...

sheepforce commented:
@ofborg build python3.pkgs.pyscf

drewrisinger commented:
Ok, I'm willing to chalk the errors up to a local build issue, I'd never seen them on my previous builds of my NUR package version.

SuperSandro2000 commented on pkgs/development/python-modules/fields/default.nix:
```diff
@@ -0,0 +1,21 @@
+    description = "Container class boilerplate killer";
+    homepage = "https://github.com/ionelmc/python-fields";
+    license = licenses.bsd2;
+    platforms = platforms.unix;
```
```suggestion
```
that is the default for buildPythonPackage, no need to repeat it.

SuperSandro2000 commented on pkgs/development/python-modules/polarizationsolver/default.nix:
```diff
@@ -0,0 +1,39 @@
+    description = "Multipole moment solver for quantum chemistry and polarisable embedding";
+    homepage = "https://gitlab.com/reinholdt/polarizationsolver";
+    license = licenses.gpl3Plus;
+    platforms = platforms.unix;
```
```suggestion
```

sheepforce committed:
python3.pkgs.polarizationsolver: init at 00424ac4


polarizationsolver: expose


polyrizationsolver: formatting


polarizationsolver: platforms


polarizationsolver: platforms


polarizationsolver: license


polarizationsolver: remove redundant platform

sheepforce committed:
python3.pkgs.fields: init at 5.0.0


fields: expose package


fields: formatting


fields: platforms


fields: platforms


fields: remove redundant platform

markuskowa commented:
I am still getting random errors from the test suite:
```
  MemoryError: Insufficient memory! MP2 memory usage 0 MB (currently available -1 MB)
  -------------------- >> begin captured stdout << ---------------------
  output file: /dev/null
  
  --------------------- >> end captured stdout << ----------------------
  
  ----------------------------------------------------------------------
  Ran 1820 tests in 2249.433s
  
  FAILED (SKIP=4, errors=2)
```
Note, that I ran this on a machine that had over 100 GB of free RAM!

markuskowa reviewed (changes requested):
We need to come up with a solution for the test suite.

markuskowa reviewed (approved):
I can't reproduce the errors anymore. Let's merge it and give it try.

markuskowa commented:
Hydra can not build it at all:
https://hydra.nixos.org/eval/1726603?filter=pyscf&compare=1726250&full=#tabs-now-fail
https://hydra.nixos.org/eval/1726676?filter=pyscf&compare=1726602&full=

sheepforce commented:
Hm this is similar to [this](https://github.com/NixOS/nixpkgs/pull/144253#pullrequestreview-803917750) problem. Neither my machines, nor ofBorg had problems with those. The number of open files can unfortunately not be increased from within the job, but is a kernel setting. Interestingly they all fail by the NWChem parser: `File "/build/source/pyscf/gto/basis/parse_nwchem.py", line 219, in search_seg`

I will take a look if I can spot something there.

sheepforce commented:
I've tried to figure out what's going on in the code there. The nwchem basis parser opens a file for each atom (or at least element) for each basis set, but it does so safely in a bracket pattern and closes them immediately. So there is no apparent reason why especially the nwchem basis set parser (with which one of the hydra jobs fails) should have too many open files. I've no idea why this is not working on Hydra and for [this](https://github.com/NixOS/nixpkgs/pull/144253#pullrequestreview-803917750) build. 

The other Hydra jobs fails with numerical issues (huge ones), which very much looks like some faulty behaviour with respect to BLAS/LAPACK - numpy interaction or something like this. I still can't reproduce any of these errors on any of my Intel machines. Hydra has some AMD Epyc instances, correct? I've also built with mkl as the blas and lapack provider for all packages (global overlay) and can also not reproduce any of the errors with MKL. 
@markuskowa , could you try building on your AMD machines with MKL for everything just once? 
I've also tried to use `blas-reference` and `lapack-reference` (this is what upstream does) but numpy refuses to build with those, due to an attribute conflict in its derivation.

If MKL also does not solve the problem I will open an issue upstream and try to get an idea from the developers.

markuskowa commented:
I can give it a try with MKL, but my suspicion is that this will not solve all the problems. It fails also on aarch64 (which we could deactivate?), where MKL. I think the problem is somewhere else: `pyscf-2.0.1` builds just fine on my internal Hydras on x86_64-linux with AMD Epyc as well as on Intel Xeon with the Openblas default. 
Looking at logs of the build failures, it looks more like resource limit of Hydra's builders (it complains about too many open files).

markuskowa commented:
FYI: building pyscf with MKL leads to the following errors. Note that they fail just above the test's threshold.
```
======================================================================
FAIL: test_c_ragf2 (test_c_agf2.KnownValues)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/source/pyscf/agf2/test/test_c_agf2.py", line 45, in test_c_ragf2
    self.assertAlmostEqual(np.max(np.absolute(vv1-vv2)), 0.0, 10)
AssertionError: 2.000888343900442e-10 != 0.0 within 10 places (2.000888343900442e-10 difference)

======================================================================
FAIL: test_c_uagf2 (test_c_agf2.KnownValues)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/build/source/pyscf/agf2/test/test_c_agf2.py", line 66, in test_c_uagf2
    self.assertAlmostEqual(np.max(np.absolute(vv1-vv2)), 0.0, 10)
AssertionError: 2.346496330574155e-10 != 0.0 within 10 places (2.346496330574155e-10 difference)

----------------------------------------------------------------------
Ran 1820 tests in 2344.443s

FAILED (SKIP=4, failures=2)
builder for '/nix/store/07675hcr0ylcm4x7nkjddcrwqba23r7q-python3.9-pyscf-2.0.1.drv' failed with exit code 1
```
